File: src/main/java/com/raychenon/hackerrank/tree/BinarySearchTree.java
package com.raychenon.hackerrank.tree;

/**
 * User: raychenon
 * Date: 29/3/19
 * https://www.hackerrank.com/challenges/is-binary-search-tree/problem
 */
public class BinarySearchTree {

    boolean checkBST(Node root) {
        return checkBSTMinMax(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    boolean checkBSTMinMax(Node root, int low, int high) {
        if (root == null) return true;
        return (root.data > low && root.data < high) &&
                checkBSTMinMax(root.left, low, root.data) &&
                checkBSTMinMax(root.right, root.data, high);
    }

}

File: src/main/kotlin/com/raychenon/kotlin/hackerrank/tree/BinarySearchTree.kt
package com.raychenon.kotlin.hackerrank.tree

/**
 * User: raychenon
 * Date: 3/4/19
 * https://www.hackerrank.com/challenges/is-binary-search-tree/problem
 */
object BinarySearchTree {

    fun checkBST(root: Node?): Boolean {
        return checkBSTMinMax(root, Int.MIN_VALUE, Int.MAX_VALUE)
    }

    private fun checkBSTMinMax(root: Node?, min: Int, max: Int): Boolean {
        // if null 
        root ?: return true

        val data: Int = root!!.data

        return (data > min && data < max) &&
                checkBSTMinMax(root.left, min, data) &&
                checkBSTMinMax(root.right, data, max)
    }
}